[BUG] bytes no longer supported

**Describe the bug**
`bytes` was serialized as integers in `v0.2.0`. Now in `v1.0.0` they're no longer supported

**To Reproduce**

The test:
```
def test_bytes(snapshot):
    snapshot.assert_match(b"hello world")
```

Result in `v1.0.0`:

```
Traceback (most recent call last):
  File "/home/ttb/ttb/test/test_tag_usb.py", line 145, in <module>
    test_bytes(s)
  File "/home/ttb/.venv/lib/python3.7/site-packages/snappiershot/snapshot/snapshot.py", line 148, in __exit__
    self._snapshot_file.write()
  File "/home/ttb/.venv/lib/python3.7/site-packages/snappiershot/snapshot/_file.py", line 142, in write
    indent=self.config.json_indentation,
  File "/home/ttb/.venv/lib/python3.7/site-packages/snappiershot/serializers/io.py", line 50, in write_json_file
    json.dump(obj, snapshot_file, cls=JsonSerializer, indent=indent, sort_keys=True)
  File "/home/ttb/.pyenv/versions/3.7.3/lib/python3.7/json/__init__.py", line 179, in dump
    for chunk in iterable:
  File "/home/ttb/.pyenv/versions/3.7.3/lib/python3.7/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/home/ttb/.pyenv/versions/3.7.3/lib/python3.7/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/home/ttb/.pyenv/versions/3.7.3/lib/python3.7/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/home/ttb/.pyenv/versions/3.7.3/lib/python3.7/json/encoder.py", line 325, in _iterencode_list
    yield from chunks
  File "/home/ttb/.pyenv/versions/3.7.3/lib/python3.7/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/home/ttb/.pyenv/versions/3.7.3/lib/python3.7/json/encoder.py", line 325, in _iterencode_list
    yield from chunks
  File "/home/ttb/.pyenv/versions/3.7.3/lib/python3.7/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/home/ttb/.venv/lib/python3.7/site-packages/snappiershot/serializers/json.py", line 80, in default
    f"Encoding for this object is not yet implemented: {value} ({type(value)})"
NotImplementedError: Encoding for this object is not yet implemented: b'hello world' (<class 'bytes'>)
```


Result in `v0.2.0`:
```
    "snappiershot_version": "0.2.0",
    "tests": {
        "test_bytes": [
            {
                "metadata": {
                    "arguments": {},
                    "test_runner_provided_name": "",
                    "update_on_next_run": false,
                    "user_provided_name": ""
                },
                "snapshots": [
                    [
                        104,
                        101,
                        108,
                        108,
                        111,
                        32,
                        119,
                        111,
                        114,
                        108,
                        100
                    ]
                ]
            }
        ],
```

But worth noting that the snapshot does fail if an update is not forced:
```
Traceback (most recent call last):
  File "/home/ttb/ttb/test/test_tag_usb.py", line 145, in <module>
    test_bytes(s)
  File "/home/ttb/ttb/test/test_tag_usb.py", line 142, in test_bytes
    snapshot.assert_match(b"hello world")
  File "/home/ttb/.venv/lib/python3.7/site-packages/snappiershot/snapshot/snapshot.py", line 90, in assert_match
    raise AssertionError(message)
AssertionError: Snapshot does not match:
- b'hello world'
+ [104, 101, 108, 108, 111, 32, 119, 111, 114, 108, 100]
------------------------------------------------------------------------------
Summary:
  > Types not equal: <class 'bytes'> != <class 'list'>
```
**Expected behavior**
`bytes` to be written to the snapshot

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Versions:**
 - Python version: 3.7.3 (pyenv)
 - Snappiershot version: `1.0.0` (catastrophic) and `0.2.0` (workable)

**Additional context**
Add any other context about the problem here.

--- a/snappiershot/serializers/json.py
+++ b/snappiershot/serializers/json.py
@@ -94,9 +94,6 @@ class JsonSerializer(json.JSONEncoder):
         if isinstance(value, PATH_TYPES):
             return self.encode_path(value)
 
-        if isinstance(value, COLLECTION_TYPES):
-            return self.encode_collection(value)
-
         raise NotImplementedError(  # pragma: no cover
             f"Encoding for this object is not yet implemented: {value} ({type(value)})"
         )

--- a/snappiershot/serializers/json.py
+++ b/snappiershot/serializers/json.py
@@ -4,7 +4,7 @@ import json
 from decimal import Decimal, DecimalTuple
 from numbers import Number
 from pathlib import Path, PurePath, PurePosixPath, PureWindowsPath
-from typing import Any, Collection, Dict, List
+from typing import Any, Collection, Dict, Iterator, List
 
 from .constants import (
     COLLECTION_TYPES,
@@ -28,6 +28,21 @@ class JsonSerializer(json.JSONEncoder):
         >>> assert json.dumps(data, cls=JsonSerializer) == '{"a": 1, "b": 2}'
     """
 
+    @classmethod
+    def _hint_tuples(cls, obj: Any) -> Any:
+        """ Convert tuples in a pre-processing step.
+
+        Extrapolated from: https://stackoverflow.com/a/15721641
+        """
+        if isinstance(obj, list):
+            return [cls._hint_tuples(item) for item in obj]
+        if isinstance(obj, COLLECTION_TYPES):
+            # Collection encoded before recursion to support sets.
+            return cls._hint_tuples(cls.encode_collection(obj))
+        if isinstance(obj, dict):
+            return {key: cls._hint_tuples(value) for key, value in obj.items()}
+        return obj
+
     def encode(self, obj: Any) -> str:
         """
         Override JSONEncoder.encode to support tuple type hinting.
@@ -38,22 +53,19 @@ class JsonSerializer(json.JSONEncoder):
 
         This method is intended to be called only by the ``json.dumps`` method.
         """
+        return super().encode(self._hint_tuples(obj))
 
-        def hint_tuples(obj_: Any) -> Any:
-            """ Convert tuples in a pre-processing step.
+    def iterencode(self, obj: Any, _one_shot: bool = False) -> Iterator[str]:
+        """
+        Override JSONEncoder.iterencode to support tuple type hinting.
 
-            Extrapolated from: https://stackoverflow.com/a/15721641
-            """
-            if isinstance(obj_, list):
-                return [hint_tuples(item) for item in obj_]
-            if isinstance(obj_, COLLECTION_TYPES):
-                # Collection encoded before recursion to support sets.
-                return hint_tuples(self.encode_collection(obj_))
-            if isinstance(obj_, dict):
-                return {key: hint_tuples(value) for key, value in obj_.items()}
-            return obj_
+        The default JSONEncoder supports both primitive types: tuples and lists. In its implementation,
+         tuples are implicitly converted to lists. To avoid this, this method allows tuples and lists
+         to be encoded as separate types.
 
-        return super().encode(hint_tuples(obj))
+        This method is intended to be called only by the ``json.dump`` method.
+        """
+        return super().iterencode(self._hint_tuples(obj), _one_shot)
 
     def default(self, value: Any) -> Any:
         """ Encode a value into a serializable object.
@@ -81,6 +93,9 @@ class JsonSerializer(json.JSONEncoder):
 
         if isinstance(value, PATH_TYPES):
             return self.encode_path(value)
+
+        if isinstance(value, COLLECTION_TYPES):
+            return self.encode_collection(value)
 
         raise NotImplementedError(  # pragma: no cover
             f"Encoding for this object is not yet implemented: {value} ({type(value)})"

--- a/tests/test_serializers/test_json.py
+++ b/tests/test_serializers/test_json.py
@@ -298,7 +298,7 @@ class TestPathEncoding:
             JsonDeserializer.decode_path(value)
 
 
-def test_round_trip():
+def test_round_trip(tmp_path: pathlib.Path):
     """ Test that a serialized and then deserialized dictionary is unchanged. """
     # Arrange
     data = {
@@ -331,12 +331,17 @@ def test_round_trip():
         "pure_posix_Path": pathlib.PurePosixPath(),
         "bytes": b"bytes",
     }
+    test_file = tmp_path / "test.json"
 
     # Act
     serialized = json.dumps(data, cls=JsonSerializer)
     deserialized = json.loads(serialized, cls=JsonDeserializer)
 
+    json.dump(data, test_file.open("w"), cls=JsonSerializer)
+    deserialized_from_file = json.load(test_file.open(), cls=JsonDeserializer)
+
     # Assert
     for key, value in deserialized.items():
         expected = data.get(key)
         assert expected == value
+    assert deserialized == deserialized_from_file

